File: backend/app/Core/Money/Currency.php
<?php

declare(strict_types=1);

namespace App\Core\Money;

use InvalidArgumentException;

/**
 * A currency the ledger can hold.
 *
 * `minorUnit` is the number of decimal places: 2 for USD, 0 for IRR, 8 for BTC.
 * This is the single fact that lets an integer amount be interpreted correctly,
 * so it travels with every Money instance.
 */
final readonly class Currency
{
    private const CATALOG = [
        'IRR' => ['ریال', 0, false, 'fiat'],
        'IRT' => ['تومان', 0, false, 'fiat'],
        'TRY' => ['₺', 2, true, 'fiat'],
        'USD' => ['$', 2, true, 'fiat'],
        'EUR' => ['€', 2, true, 'fiat'],
        'AED' => ['د.إ', 2, false, 'fiat'],
        'BTC' => ['₿', 8, true, 'crypto'],
        'ETH' => ['Ξ', 8, true, 'crypto'],
        'USDT' => ['₮', 2, true, 'crypto'],
        'XAU' => ['طلا', 4, false, 'metal'],
    ];

    private function __construct(
        public string $code,
        public string $symbol,
        public int $minorUnit,
        public bool $symbolLeading,
        public string $type,
    ) {}

    public static function of(string $code): self
    {
        $code = strtoupper(trim($code));

        if (! isset(self::CATALOG[$code])) {
            throw new InvalidArgumentException("Unknown currency [{$code}].");
        }

        [$symbol, $minorUnit, $symbolLeading, $type] = self::CATALOG[$code];

        return new self($code, $symbol, $minorUnit, $symbolLeading, $type);
    }

    public static function isSupported(string $code): bool
    {
        return isset(self::CATALOG[strtoupper(trim($code))]);
    }

    /** @return list<string> */
    public static function codes(): array
    {
        return array_keys(self::CATALOG);
    }

    /** @return list<array{code:string,symbol:string,minor_unit:int,symbol_leading:bool,type:string}> */
    public static function catalog(): array
    {
        $rows = [];

        foreach (self::CATALOG as $code => [$symbol, $minorUnit, $symbolLeading, $type]) {
            $rows[] = [
                'code' => $code,
                'symbol' => $symbol,
                'minor_unit' => $minorUnit,
                'symbol_leading' => $symbolLeading,
                'type' => $type,
            ];
        }

        return $rows;
    }

    /** 10 ** minorUnit — the divisor between minor and major units. */
    public function factor(): int
    {
        return 10 ** $this->minorUnit;
    }

    public function equals(self $other): bool
    {
        return $this->code === $other->code;
    }

    public function __toString(): string
    {
        return $this->code;
    }
}

File: app/lib/core/money/currency.dart
/// A currency the app can hold. `minorUnit` is the number of decimal places:
/// 2 for USD, 0 for IRR, 8 for BTC.
final class Currency {
  const Currency({
    required this.code,
    required this.symbol,
    required this.minorUnit,
    this.symbolLeading = true,
  });

  final String code;
  final String symbol;
  final int minorUnit;

  /// Whether the symbol goes before the number in LTR locales.
  final bool symbolLeading;

  static const irr = Currency(code: 'IRR', symbol: 'ریال', minorUnit: 0, symbolLeading: false);
  static const toman = Currency(code: 'IRT', symbol: 'تومان', minorUnit: 0, symbolLeading: false);
  static const try_ = Currency(code: 'TRY', symbol: '₺', minorUnit: 2);
  static const usd = Currency(code: 'USD', symbol: '\$', minorUnit: 2);
  static const eur = Currency(code: 'EUR', symbol: '€', minorUnit: 2);
  static const aed = Currency(code: 'AED', symbol: 'د.إ', minorUnit: 2, symbolLeading: false);
  static const btc = Currency(code: 'BTC', symbol: '₿', minorUnit: 8);
  static const eth = Currency(code: 'ETH', symbol: 'Ξ', minorUnit: 8);
  static const usdt = Currency(code: 'USDT', symbol: '₮', minorUnit: 2);

  static const all = <Currency>[irr, toman, try_, usd, eur, aed, btc, eth, usdt];

  static Currency? byCode(String code) {
    for (final currency in all) {
      if (currency.code == code.toUpperCase()) return currency;
    }
    return null;
  }

  @override
  bool operator ==(Object other) => other is Currency && other.code == code;

  @override
  int get hashCode => code.hashCode;

  @override
  String toString() => code;
}
